## House Price Prediction Using Machine Learning

## Project Objective

The objective of this project is to build a machine learning regression model
that predicts house prices based on various property and demographic features.

The project includes data preprocessing, exploratory data analysis (EDA),
feature analysis, model training, and evaluation using regression metrics.

## Import Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

## Load Dataset

In [2]:
df = pd.read_csv("data/raw/housing.csv")

In [3]:
df.head()

,Avg. Area Income,Avg. Area House Age,Avg. Area Number of Rooms,Avg. Area Number of Bedrooms,Area Population,Price,Address
0,79545.45857,5.682861,7.009188,4.09,23086.80050,1.059034e+06,"208 Michael Ferry Apt. 674\nLaurabury, NE 3701..."
1,79248.64245,6.002900,6.730821,3.09,40173.07217,1.505891e+06,"188 Johnson Views Suite 079\nLake Kathleen, CA..."
2,61287.06718,5.865890,8.512727,5.13,36882.15940,1.058988e+06,"9127 Elizabeth Stravenue\nDanieltown, WI 06482..."
3,63345.24005,7.188236,5.586729,3.26,34310.24283,1.260617e+06,USS Barnett\nFPO AP 44820
4,59982.19723,5.040555,7.839388,4.23,26354.10947,6.309435e+05,USNS Raymond\nFPO AE 09386


In [4]:
df.shape

(5000, 7)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Avg. Area Income              5000 non-null   float64
 1   Avg. Area House Age           5000 non-null   float64
 2   Avg. Area Number of Rooms     5000 non-null   float64
 3   Avg. Area Number of Bedrooms  5000 non-null   float64
 4   Area Population               5000 non-null   float64
 5   Price                         5000 non-null   float64
 6   Address                       5000 non-null   object 
dtypes: float64(6), object(1)
memory usage: 273.6+ KB


In [6]:
df.describe()

,Avg. Area Income,Avg. Area House Age,Avg. Area Number of Rooms,Avg. Area Number of Bedrooms,Area Population,Price
count,5000.000000,5000.000000,5000.000000,5000.000000,5000.000000,5.000000e+03
mean,68583.108984,5.977222,6.987792,3.981330,36163.516039,1.232073e+06
std,10657.991214,0.991456,1.005833,1.234137,9925.650114,3.531176e+05
min,17796.631190,2.644304,3.236194,2.000000,172.610686,1.593866e+04
25%,61480.562390,5.322283,6.299250,3.140000,29403.928700,9.975771e+05
50%,68804.286405,5.970429,7.002902,4.050000,36199.406690,1.232669e+06
75%,75783.338665,6.650808,7.665871,4.490000,42861.290770,1.471210e+06
max,107701.748400,9.519088,10.759588,6.500000,69621.713380,2.469066e+06


In [7]:
df.isnull().sum()

Avg. Area Income                0
Avg. Area House Age             0
Avg. Area Number of Rooms       0
Avg. Area Number of Bedrooms    0
Area Population                 0
Price                           0
Address                         0
dtype: int64

In [9]:
df.duplicated().sum()

0

In [10]:
df.columns

Index(['Avg. Area Income', 'Avg. Area House Age', 'Avg. Area Number of Rooms',
       'Avg. Area Number of Bedrooms', 'Area Population', 'Price', 'Address'],
      dtype='object')

In [11]:
df.select_dtypes(include="object").columns

Index(['Address'], dtype='object')

In [15]:
clean_df = df.drop(columns=["Address"])

## Save Cleaned Dataset

In [16]:
clean_df.to_csv("data/processed/clean_housing.csv", index=False)

##  Train-Test Split

In [17]:
from sklearn.model_selection import train_test_split

X = clean_df.drop("Price", axis=1)
y = clean_df["Price"]

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

print("Training features:", X_train.shape)
print("Testing features:", X_test.shape)
print("Training target:", y_train.shape)
print("Testing target:", y_test.shape)

Training features: (4000, 5)
Testing features: (1000, 5)
Training target: (4000,)
Testing target: (1000,)


## Feature Scaling


In [18]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

print("Scaled training data shape:", X_train_scaled.shape)
print("Scaled testing data shape:", X_test_scaled.shape)

Scaled training data shape: (4000, 5)
Scaled testing data shape: (1000, 5)


## Model Building


In [19]:
from sklearn.linear_model import LinearRegression

# Initialize the model
linear_model = LinearRegression()

# Train the model
linear_model.fit(X_train_scaled, y_train)

print("Linear Regression model trained successfully.")

Linear Regression model trained successfully.


## Model Prediction

In [20]:
# Generate predictions on the test data
y_pred_lr = linear_model.predict(X_test_scaled)

print("First 10 Actual Prices:")
print(y_test.head(10).values)

print("\nFirst 10 Predicted Prices:")
print(y_pred_lr[:10])

First 10 Actual Prices:
[1339096.077 1251794.179 1340094.966 1431507.623 1042373.524 1555320.5
 1250882.292 1039380.722  832475.189 1420648.281]

First 10 Predicted Prices:
[1308587.92718477 1237037.22950292 1243429.34023015 1228900.21368723
 1063320.90731644 1544058.05050119 1094774.70494245  833284.72338221
  788412.85579589 1469714.86611301]


## Model Evaluation

In [21]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

mae_lr = mean_absolute_error(y_test, y_pred_lr)
mse_lr = mean_squared_error(y_test, y_pred_lr)
rmse_lr = np.sqrt(mse_lr)
r2_lr = r2_score(y_test, y_pred_lr)

print("Linear Regression Performance:")
print(f"MAE  : {mae_lr:.2f}")
print(f"MSE  : {mse_lr:.2f}")
print(f"RMSE : {rmse_lr:.2f}")
print(f"R²   : {r2_lr:.4f}")

Linear Regression Performance:
MAE  : 80879.10
MSE  : 10089009299.50
RMSE : 100444.06
R²   : 0.9180


## Random Forest Regressor


In [22]:
from sklearn.ensemble import RandomForestRegressor

# Initialize the model
rf_model = RandomForestRegressor(
    n_estimators=100,
    random_state=42
)

# Train the model
rf_model.fit(X_train, y_train)

print("Random Forest Regressor model trained successfully.")

Random Forest Regressor model trained successfully.


## Random Forest Prediction


In [23]:
# Generate predictions
y_pred_rf = rf_model.predict(X_test)

print("First 10 Actual Prices:")
print(y_test.head(10).values)

print("\nFirst 10 Predicted Prices:")
print(y_pred_rf[:10])

First 10 Actual Prices:
[1339096.077 1251794.179 1340094.966 1431507.623 1042373.524 1555320.5
 1250882.292 1039380.722  832475.189 1420648.281]

First 10 Predicted Prices:
[1319175.46735  1242701.582115 1270790.952625 1218569.277491
 1195450.115877 1529217.74386  1065196.582172  848999.453174
  888152.68533  1383945.2183  ]


##  Random Forest Model Evaluation


In [24]:
mae_rf = mean_absolute_error(y_test, y_pred_rf)
mse_rf = mean_squared_error(y_test, y_pred_rf)
rmse_rf = np.sqrt(mse_rf)
r2_rf = r2_score(y_test, y_pred_rf)

print("Random Forest Performance:")
print(f"MAE  : {mae_rf:.2f}")
print(f"MSE  : {mse_rf:.2f}")
print(f"RMSE : {rmse_rf:.2f}")
print(f"R²   : {r2_rf:.4f}")

Random Forest Performance:
MAE  : 94511.34
MSE  : 14391497208.58
RMSE : 119964.57
R²   : 0.8830


## Model Comparison

In [25]:
import pandas as pd

model_comparison = pd.DataFrame({
    "Model": ["Linear Regression", "Random Forest"],
    "MAE": [mae_lr, mae_rf],
    "MSE": [mse_lr, mse_rf],
    "RMSE": [rmse_lr, rmse_rf],
    "R2 Score": [r2_lr, r2_rf]
})

model_comparison

,Model,MAE,MSE,RMSE,R2 Score
0,Linear Regression,80879.097222,1.008901e+10,100444.060549,0.917997
1,Random Forest,94511.340898,1.439150e+10,119964.566471,0.883027


## Best Model Selection


In [28]:
best_model = linear_model

print("Best Model: Linear Regression")
print(f"R² Score: {r2_lr:.4f}")
print(f"RMSE: {rmse_lr:.2f}")

Best Model: Linear Regression
R² Score: 0.9180
RMSE: 100444.06


## Save the Final Model

In [39]:
import joblib

joblib.dump(best_model, "models/linear_regression_model.pkl")
joblib.dump(scaler, "models/standard_scaler.pkl")

print("Model and scaler saved successfully.")

Model and scaler saved successfully.


## Conclusion

In this project, a machine learning approach was used to predict house prices based on various housing-related features.

The dataset was cleaned and analyzed through exploratory data analysis, followed by train-test splitting and feature scaling. Two regression models, Linear Regression and Random Forest Regressor, were trained and evaluated.

Based on the evaluation results, Linear Regression performed better than Random Forest Regressor, achieving an R² score of 0.9180 on the test dataset.

The final Linear Regression model and StandardScaler were saved using Joblib for future predictions.

### Final Model Performance

- **Model:** Linear Regression
- **R² Score:** 0.9180
- **RMSE:** 100,444.06